source/MaterialXCore/Document.cpp

```diff
@@ -171,46 +171,127 @@ void Document::initialize()
 }
 
 NodeDefPtr Document::addNodeDefFromGraph(const NodeGraphPtr nodeGraph, const string& nodeDefName, const string& node,
-                                         const string& version, bool isDefaultVersion, const string& group, const string& newGraphName)
+    const string& version, bool isDefaultVersion, const string& group, const string& newGraphName)
 {
-    if (getNodeDef(nodeDefName))
+    DefinitionOptions options;
+    options.categoryString = node;
+    options.compoundGraph = nodeGraph;
+    options.docString = EMPTY_STRING;
+    options.isDefaultVersion = isDefaultVersion;
+    options.namespaceString = EMPTY_STRING;
+    options.nodeGroupString = group;
+    options.useNamespace = false;
+    options.useVersion = !version.empty();
+    options.versionString = version;
+
+    options.newNodeDefName = nodeDefName;
+    options.newNodeGraphName = newGraphName;
+
+    return createDefinition(options);
+}
+
+NodeDefPtr Document::createDefinition(DefinitionOptions& options)
```

Comment:
Ah, I think I was imagining exactly the opposite approach, where the new version of `Document::createDefinition` would have no inputs for features that can be set just as easily by the client (e.g. doc string, node group string, version string, default version).  I wasn't imagining that we'd bundle all of these inputs into a new `DefinitionOptions` structure, but that the inputs wouldn't exist at all!

What are you thinking of as the motivation for providing these inputs as arguments of `Document::createDefinition`, rather than just letting the client set them after they create their NodeDef from the graph?

Reply:
> 

The main impetus is to:
1. Make this an atomic operation where you don't need, and don't want users to go back and change
the signature. Conversely avoid the inherent pitfalls of modifying a definition after it's been created without forethought.
2. Make it so that all relevant arguments are up-front to avoid not noticing important arguments which affect behaviour and signature. 

For example, category is required, nodegroup affects behaviour, namespace needs to be set on 2 elements and this and version should never really be changed after the fact. 

I think the only attribute which can be trivially changed is the doc string.